compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -210,19 +226,49 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
         NestedVisitorMap::None
     }
 
+    fn visit_arm(&mut self, arm: &'tcx Arm<'tcx>) {
+        if arm.guard.is_some() {
+            self.nested_scope_of_guards.push(<_>::default());
+            self.arm_has_guard = true;
```

Comment:
It seems like we'd only want to set this while visiting the guard if expression, not the pattern.